[expo-camera][ios] Use CoreMotion to detect device orientation

chalenascholl opened the pull request:
# Why

Expo-camera doesn't allow landscape photos when the screen orientation is locked or if the React native app orientation is locked. This is contrary to how most cameras work (including apple's native camera), which allow you to take landscape photos despite orientation being locked to portrait.

Closes https://github.com/expo/expo/issues/21938


# How

The reason this happens is because `expo-camera` [relies on UIDeviceOrientation](https://github.com/expo/expo/blob/6085491bf6236a1aa82f96616ea31bbf4eda06ae/packages/expo-camera/ios/EXCamera/EXCamera.m#L357) to determine what type of a photo to take (portrait or landscape). It then sets the AVVideoOrientation to pass to Apple's `takePicture` or `startRecordingToOutputFileURL` function based on that. However, if the orientation is locked, the device orientation is always given as portrait by Apple which means the picture is always taken in portrait.

The only workaround I am aware of is to use the `CoreMotion` framework to check where the device is in space and then use that to determine what type of photo should be taken (portrait or landscape). This was also how the react-native-camera [package worked](https://github.com/react-native-camera/react-native-camera/blob/7cdfb6ec127decf1ac5686635afb7f07557becc5/ios/RCT/RCTSensorOrientationChecker.m#L72)

- I've added a property (ios only) `responsiveOrientationWhenOrientationLocked' to determine whether or not we should use the physical orientation of the device to determine the capture orientation when orientation is locked. This way, we won't be breaking any users of the package who rely on the existing behavior.
- I've also added a callback that applies when this property is set to true. Since `responsiveOrientationWhenOrientationLocked` means that landscape photos can be taken when the app is locked to portrait or when the device is locked to portrait, the screen itself will no longer rotate. The callback provides the user with the ability to rotate icons/buttons when the camera capture orientation changes (also what the native iOS camera does)

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

schalena committed:
update expo camera to use coremotion to detect device orientation

schalena committed:
update changelog for expo-camera

chalenascholl committed:
Update packages/expo-camera/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

lukmccall reviewed (changes requested):
Thanks for your contribution. While I find the concept of setting that as the standard appealing, it's important to consider that someone may be reliant on the existing behavior. Therefore, we should introduce some form of toggle or preference.

schalena committed:
add type to allow responsive camera orientation when screen orientation locked on ios

chalenascholl commented:
I added a property for iOS, to optionally set whether or not responsive camera orientation should be allowed if screen orientation is locked. 

I also updated the PR description to add that landscape photos are also not allowed if the React Native app is locked to portrait orientation.

chalenascholl commented on packages/expo-camera/ios/EXCamera/EXCamera.m:
```diff
@@ -709,6 +717,13 @@ - (void)startSession
       });
     }]];
 
+    [self.motionManager startAccelerometerUpdatesToQueue:[NSOperationQueue new]
```
Perhaps it would be better to start and stop accelerometer updates to the queue only when the `responsiveOrientationWhenOrientationLocked` property is changed?

schalena committed:
start and stop accelerometer updates based on responsive orientation property value

schalena committed:
remove semicolon in swift change

schalena committed:
fix spacing

schalena committed:
add onResponsiveOrientationChanged callback property for ios

lukmccall commented on packages/expo-camera/src/Camera.types.ts:
```diff
@@ -105,6 +105,13 @@ export enum VideoQuality {
   '4:3' = '4:3',
 }
 
+export enum CameraOrientation {
```
I think those enums should be defined as ints instead of strings 🤔 
```ts
export enum CameraOrientation {
  portrait = 1,
  portraitUpsideDown = 2,
  landscapeLeft = 3,
  landscapeRight = 4
}
```
By doing that someone can write code like this:
```ts
onResponsiveOrientationChanged={({orientation}) => {
  if (orientation === CameraOrientation.portrait) {
  	// do something 
  }
}}
```
It should be enough for that case

chalenascholl replied:
Ah yes, good idea. Thanks

schalena committed:
update enum for cameraorientation to numbers

lukmccall reviewed (approved):
It appears that the JavaScript files haven't been pre-built yet. Kindly complete this step and once done, I will merge your PR.

schalena committed:
fix formatting of code
